### PR TITLE
Exception thrown trying to retrieve CourseStudent: IllegalBlockSizeException #6562

### LINKED ISSUE #6562: Exception thrown trying to retrieve CourseStudent: IllegalBlockSizeException
Found in the log of the live server V5.94

```
Error Message: teammates.common.util.Logger severe: Exception thrown trying to retrieve CourseStudent

java.lang.RuntimeException: javax.crypto.IllegalBlockSizeException: Input length must be multiple of 16 when decrypting with padded cipher
at teammates.common.util.StringHelper.decrypt(StringHelper.java:147)
at teammates.storage.api.StudentsDb.getStudentForRegistrationKey(StudentsDb.java:190)
at teammates.logic.core.StudentsLogic.getStudentForRegistrationKey(StudentsLogic.java:93)
at teammates.logic.api.Logic.getStudentForRegistrationKey(Logic.java:815)
at teammates.ui.controller.Action.authenticateNotLoggedInUser(Action.java:196)
at teammates.ui.controller.Action.authenticateAndGetActualUser(Action.java:144)
at teammates.ui.controller.Action.authenticateUser(Action.java:125)
at teammates.ui.controller.Action.init(Action.java:89)
at teammates.ui.controller.ActionFactory.getAction(ActionFactory.java:150)
at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:64)
Caused by: javax.crypto.IllegalBlockSizeException: Input length must be multiple of 16 when decrypting with padded cipher
at com.sun.crypto.provider.CipherCore.doFinal(CipherCore.java:750)
at com.sun.crypto.provider.CipherCore.doFinal(CipherCore.java:676)
at com.sun.crypto.provider.AESCipher.engineDoFinal(AESCipher.java:313)
at javax.crypto.Cipher.doFinal(Cipher.java:2131)
at teammates.common.util.StringHelper.decrypt(StringHelper.java:144)
```

### PR DESCRIPTION
**Pull request by Neurrone**
Fixes #6562 

Outline of solution:

* Handle different exception cases in StringHelper's decrypt method:
	* rethrow IllegalBlockSizeException or BadPaddingException as an InvalidParametersException
	* otherwise, rethrow as a RuntimeException (in this case, something horrible like failure to initialize Cipher happened)
* Modify client methods to handle the InvalidParametersException appropriately
* Add a constructor that accepts a Throwable in TeammatesException, InvalidParametersException and EntityDoesNotExistException so that PMD doesn't complain about stack trace info being lost when rethrowing exceptions as different types.
* Add tests for new decryption behaviour

### CODE REVIEW

**Comment by Neurrone** [comment]
The invalid block size exception is thrown if the ciphertext is invalid.

StringHelper's decrypt method currently rethrows any exception as a RuntimeException, which makes it harder to know what went wrong. It'd be good to distinguish between invalid ciphertext, or some more serious error (e.g Cipher cannot be initialized).

Proposed solution:

* rethrow IllegalBlockSizeException or BadPaddingException as an InvalidParametersException if invalid ciphertext
* otherwise, rethrow as a RuntimeException

However, this requires clients to handle the possibility.

The alternative is to leave the decrypt method unchanged, catch any exception that occurs while decrypting and assume it is invalid input, not anything more serious.

**Comment by junhaoyap** [comment]
Please assign someone and let them know it's ready for review when it is ready, I have set it to ongoing in the mean time

**Comment by Neurrone** [comment]
Ready for review - proposed solution implemented.

File: src/main/java/teammates/ui/controller/StudentProfilePictureAction.java [hunk 1]
```diff
@@ -42,8 +43,16 @@ private ActionResult handleRequestWithBlobKey() {
...
-        String email = getStudentEmailFromRequest();
-        String courseId = getCourseIdFromRequest();
+        String email;
+        String courseId;
+        try {
+            email = getStudentEmailFromRequest();
+            courseId = getCourseIdFromRequest();
+        } catch (InvalidParametersException e) {
+            // potential PMD bug? No stack trace info is lost here
+            throw new EntityDoesNotExistException(e.getMessage()); //NOPMD
```
Comment (ablyx) [inline src/main/java/teammates/ui/controller/StudentProfilePictureAction.java]: Is this a `EntityDoesNotExistException`? should there be a log message instead/as well?

File: src/main/java/teammates/storage/api/StudentsDb.java [hunk 2]
```diff
@@ -180,17 +180,17 @@ public StudentAttributes getStudentForGoogleId(String courseId, String googleId)
...
-            String originalKey = StringHelper.decrypt(registrationKey.trim());
-            CourseStudent courseStudent = getCourseStudentEntityForRegistrationKey(originalKey);
+            String decryptedKey = StringHelper.decrypt(encryptedRegistrationKey.trim());
+            CourseStudent courseStudent = getCourseStudentEntityForRegistrationKey(decryptedKey);
             if (courseStudent == null) {
                 return null;
             }
             return new StudentAttributes(courseStudent);
+        } catch (InvalidParametersException e) {
+            return null; // invalid registration key cannot be decrypted
```
Comment (ablyx) [inline src/main/java/teammates/storage/api/StudentsDb.java]: maybe log this?

File: src/main/java/teammates/storage/api/InstructorsDb.java [hunk 3]
```diff
@@ -178,10 +178,14 @@ public InstructorAttributes getInstructorForGoogleId(String courseId, String goo
...
     public InstructorAttributes getInstructorForRegistrationKey(String encryptedKey) {
-        
         Assumption.assertNotNull(Const.StatusCodes.DBLEVEL_NULL_INPUT, encryptedKey);
         
-        String decryptedKey = StringHelper.decrypt(encryptedKey.trim());
+        String decryptedKey;
+        try {
+            decryptedKey = StringHelper.decrypt(encryptedKey.trim());
+        } catch (InvalidParametersException e) {
+            return null;
```
Comment (ablyx) [inline src/main/java/teammates/storage/api/InstructorsDb.java]: maybe log this?

File: src/main/java/teammates/storage/api/StudentsDb.java (hunk 2 above)
Comment (Neurrone) [inline src/main/java/teammates/storage/api/StudentsDb.java]: The comment for this function indicates that callers should expect null, so don't think its an exceptional case.

File: src/main/java/teammates/storage/api/InstructorsDb.java (hunk 3 above)
Comment (Neurrone) [inline src/main/java/teammates/storage/api/InstructorsDb.java]: Similarly as well.

File: src/main/java/teammates/ui/controller/StudentProfilePictureAction.java (hunk 1 above)
Comment (Neurrone) [inline src/main/java/teammates/ui/controller/StudentProfilePictureAction.java]: Exceptions are caught further up at the call stack and logged. Maybe what I could do instead is to make the exception message more descriptive. I.e, prepend something like "Message could not be decrypted:" followed by java's original message.

Comment (Neurrone) [inline src/main/java/teammates/ui/controller/StudentProfilePictureAction.java]: Sorry I see what you mean now, will log this.

**Comment by Neurrone** [comment]
@ablyx changes made.

**Comment by Neurrone** [comment]
@chowyb changes made.

File: src/main/java/teammates/common/util/StringHelper.java [hunk 4]
```diff
@@ -134,13 +140,25 @@ public static String encrypt(String value) {
         }
     }
 
-    public static String decrypt(String message) {
+    /*
+     * Decrypts the supplied string.
+     *
+     * @param message the ciphertext as a hexidecimal string
```
Comment (chowyb) [inline src/main/java/teammates/common/util/StringHelper.java]: Use `hexadecimal` instead of `hexidecimal`.

**Comment by Neurrone** [comment]
@chowyb typo fixed.

File: src/main/java/teammates/common/util/StringHelper.java [hunk 5]
```diff
@@ -134,13 +140,25 @@ public static String encrypt(String value) {
...
         try {
             SecretKeySpec sks = new SecretKeySpec(hexStringToByteArray(Config.ENCRYPTION_KEY), "AES");
             Cipher cipher = Cipher.getInstance("AES");
             cipher.init(Cipher.DECRYPT_MODE, sks);
             byte[] decrypted = cipher.doFinal(hexStringToByteArray(message));
             return new String(decrypted);
+        } catch (NumberFormatException | IllegalBlockSizeException | BadPaddingException e) {
+            log.warning("Attempted to decrypt invalid ciphertext: " + message);
+            // potential PMD bug? No stack trace info is lost here
+            throw new InvalidParametersException(e.getMessage()); //NOPMD
```
Comment (damithc) [inline src/main/java/teammates/common/util/StringHelper.java]: I think `e.getMessage()` doesn't contain stack trace. Is it possible to do `InvalidParametersException(e)` instead?

File: src/test/java/teammates/test/cases/util/StringHelperTest.java [hunk 6]
```diff
@@ -115,12 +116,19 @@ public void testToString() {
...
 
         decrptedMsg = StringHelper.decrypt(StringHelper.encrypt(msg));
         assertEquals(msg, decrptedMsg);
+
+        try {
+            StringHelper.decrypt("ABC");
+            signalFailureToDetectException("Decryption of invalid ciphertext should throw");
+        } catch (InvalidParametersException e) {
+            ignoreExpectedException();
+        }
```
Comment (damithc) [inline src/test/java/teammates/test/cases/util/StringHelperTest.java]: Perhaps use ExpectedException technique here?
http://howtodoinjava.com/testng/testng-expected-exception-and-expected-message-tutorial/

File: src/main/java/teammates/common/util/StringHelper.java (hunk 5 above)
Comment (Neurrone) [inline src/main/java/teammates/common/util/StringHelper.java]: There isn't a constructor defined that takes in Exception. Should I create one?

**Comment by Neurrone** [comment]
@damithc change for expected exception technique made.

File: src/test/java/teammates/test/cases/util/StringHelperTest.java [hunk 7]
```diff
@@ -115,14 +116,19 @@ public void testToString() {
...
-    public void testKeyEncryption() {
+    public void testKeyEncryption() throws InvalidParametersException {
         String msg = "Test decryption";
         String decrptedMsg;
 
         decrptedMsg = StringHelper.decrypt(StringHelper.encrypt(msg));
         assertEquals(msg, decrptedMsg);
     }
 
+    @Test(expectedExceptions = { InvalidParametersException .class })
```
Comment (damithc) [inline src/test/java/teammates/test/cases/util/StringHelperTest.java]: There is a space between `InvalidParametersException` and `.class`.
Is that intended?

Also, should there be a space after `{` and before `}`?

File: src/test/java/teammates/test/cases/util/StringHelperTest.java [hunk 8]
```diff
@@ -124,7 +124,7 @@ public void testKeyEncryption() throws InvalidParametersException {
         assertEquals(msg, decrptedMsg);
     }
 
-    @Test(expectedExceptions = { InvalidParametersException .class })
+    @Test(expectedExceptions = {InvalidParametersException .class})
```
Comment (ablyx) [inline src/test/java/teammates/test/cases/util/StringHelperTest.java]: There shouldn't be a space between InvalidParametersException and .class

**Comment by Neurrone** [comment]
@ablyx done.

File: src/test/java/teammates/test/cases/util/StringHelperTest.java [hunk 9]
```diff
@@ -115,14 +116,19 @@ public void testToString() {
...
         String msg = "Test decryption";
         String decrptedMsg;
 
         decrptedMsg = StringHelper.decrypt(StringHelper.encrypt(msg));
         assertEquals(msg, decrptedMsg);
     }
 
+    @Test(expectedExceptions = {InvalidParametersException.class})
+    public void testEncryptingInvalidPlaintextThrowsException() throws InvalidParametersException {
+        StringHelper.decrypt("ABC");
```
Comment (damithc) [inline src/test/java/teammates/test/cases/util/StringHelperTest.java]: Any chance of using equivalence partitions and boundary values to come up with better test cases? Or 'a random string' is the only test case that we can use?

File: src/main/java/teammates/common/util/StringHelper.java (hunk 5 above)
Comment (damithc) [inline src/main/java/teammates/common/util/StringHelper.java]: Sorry, I missed that question. Yes, we can try that. But can be done as a separate PR if you wish.

**Comment by Neurrone** [comment]
@damithc changes made. Changed to using previous try-catch block since this way would be more concise with multiple cases.

File: src/test/java/teammates/test/cases/util/StringHelperTest.java [hunk 10]
```diff
@@ -124,6 +125,23 @@ public void testKeyEncryption() {
     }
 
     @Test
+    public void testDecryptingInvalidCiphertextThrowsException() {
+        String invalidHexString = "GHI";
+        // each hex digit is 4 bits
+        String ciphertextLength120 = "AAAAAAAAAABBBBBBBBBBCCCCCCCCCC";
+        String ciphertextLength136 = ciphertextLength120 + "1234";
+        String[] invalidCiphertexts = {invalidHexString, ciphertextLength120, ciphertextLength136};
```
Comment (damithc) [inline src/test/java/teammates/test/cases/util/StringHelperTest.java]: @Neurrone Can explain a bit more on why each of these are invalid?

**Comment by Neurrone** [comment]
@damithc The decrypt function converts a hex string into an array of bytes before they are decrypted, so e.g AF is the byte  10101111. So non-hex strings should fail to decrypt. 

AES requires the length of data to be multiples of 128 bits. So it checks if it fails on inputs of 120 and 136 bits.

File: src/test/java/teammates/test/cases/util/StringHelperTest.java (hunk 10 above)
Comment (Neurrone) [inline src/test/java/teammates/test/cases/util/StringHelperTest.java]: @damithc The decrypt function converts a hex string into an array of bytes before they are decrypted, so e.g AF is the byte  10101111. So non-hex strings should fail to decrypt. 

AES requires the length of data to be multiples of 128 bits. So it checks if it fails on inputs of 120 and 136 bits.

Comment (damithc) [inline src/test/java/teammates/test/cases/util/StringHelperTest.java]: Got it. Can you add that as a comment?

Comment (Neurrone) [inline src/test/java/teammates/test/cases/util/StringHelperTest.java]: Added.

**Pull request merged**